File: lib/models/card_instance.dart
/// Represents a specific instance of a card in a game.
/// Each physical card in a deck has its own CardInstance, even if multiple
/// cards share the same cardId (duplicates of the same card type).
class CardInstance {
  final String instanceId;
  final String cardId;

  const CardInstance({
    required this.instanceId,
    required this.cardId,
  });

  factory CardInstance.fromJson(Map<String, dynamic> json) {
    return CardInstance(
      instanceId: json['instanceId']?.toString() ?? '',
      cardId: json['cardId']?.toString() ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'instanceId': instanceId,
      'cardId': cardId,
    };
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is CardInstance && other.instanceId == instanceId;
  }

  @override
  int get hashCode => instanceId.hashCode;

  @override
  String toString() {
    return 'CardInstance(instanceId: $instanceId, cardId: $cardId)';
  }
}

File: backend/internal/domain/card_instance.go
package domain

import (
	"fmt"
	"github.com/google/uuid"
)

// CardInstanceID uniquely identifies a specific card instance in a game.
// This is different from CardID which identifies the card type/definition.
type CardInstanceID string

// NewCardInstanceID generates a new unique card instance ID.
func NewCardInstanceID() CardInstanceID {
	return CardInstanceID(uuid.New().String())
}

// CardInstance represents a specific instance of a card in a game.
// Each physical card in a deck has its own CardInstance, even if multiple
// cards share the same CardID (duplicates of the same card type).
type CardInstance struct {
	InstanceID CardInstanceID `json:"instanceId"`
	CardID     CardID         `json:"cardId"`
}

// NewCardInstance creates a new card instance for a given card type.
func NewCardInstance(cardID CardID) CardInstance {
	return CardInstance{
		InstanceID: NewCardInstanceID(),
		CardID:     cardID,
	}
}

// String returns a string representation of the card instance.
func (ci CardInstance) String() string {
	return fmt.Sprintf("CardInstance{ID: %s, Card: %s}", ci.InstanceID, ci.CardID)
}